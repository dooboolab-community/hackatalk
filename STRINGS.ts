import * as Localization from 'expo-localization';

import i18n from 'i18n-js';

const en = {
  LOADING: 'Loading...',
  HELLO: 'Hello',
  LOGIN: 'Login',
  EMAIL: 'Email',
  PASSWORD: 'Password',
  SIGN_UP: 'Sign Up',
  LOGOUT: 'Logout',
  FORGOT_PW: 'Forgot password?',
  FIND_PW: 'Find Password',
  NAME: 'name',
  STATUS_MSG: 'Status message',
  REGISTER: 'Register',
  SEND_LINK: 'Send Link',
  FRIEND: 'Friend',
  MESSAGE: 'Message',
  MY_PROFILE: 'My profile',
  UPDATE: 'Update',
  NO_CONTENT: 'No content',
  ADD_FRIEND: 'Add',
  DELETE_FRIEND: 'Delete',
  SEARCH_USER: 'Search User',
  CHAT: 'Chat',
  SEND: 'Send',
  WRITE_MESSAGE: 'Write message...',
  ERROR: 'Error',
  FRIEND_ADDED: 'Added to your friend.',
  FRIEND_ALREADY_ADDED: 'Already your friend.',
  SETTING: 'Setting',
  SIGN_IN_WITH_GOOGLE: 'Sign in with Google',
  SIGN_IN_WITH_FACEBOOK: 'Sign in with Facebook',
};

const ko = {
  LOADING: '로딩중...',
  HELLO: '안녕하세요',
  LOGIN: '로그인',
  LOGOUT: '로그아웃',
  EMAIL: '이메일',
  PASSWORD: '패스워드',
  SIGN_UP: '회원가입',
  FORGOT_PW: '비밀번호를 잊어버리셨나요?',
  FIND_PW: '비밀번호 찾기',
  NAME: '이름',
  STATUS_MSG: '상태메세지',
  REGISTER: '가입',
  SEND_LINK: '링크 보내기',
  FRIEND: '친구',
  MESSAGE: '메세지',
  MY_PROFILE: '나의 프로필',
  UPDATE: '수정',
  NO_CONTENT: '컨텐츠가 없습니다',
  ADD_FRIEND: '친구추가',
  DELETE_FRIEND: '친구삭제',
  SEARCH_USER: '사용자 검색',
  CHAT: '채팅',
  SEND: '보내기',
  WRITE_MESSAGE: '메세지를 입력해주세요…',
  ERROR: '에러',
  FRIEND_ADDED: '친구로 등록되었습니다.',
  FRIEND_ALREADY_ADDED: '이미 친구입니다.',
  SETTING: '설정',
  SIGN_IN_WITH_GOOGLE: '구글 계정으로 시작하기',
  SIGN_IN_WITH_FACEBOOK: '페이스북 계정으로 시작하기',
};

i18n.fallbacks = true;
i18n.translations = { en, ko };
i18n.locale = Localization.locale;

export const getString = (param: string, mapObj?: object): string => {
  if (mapObj) {
    i18n.t(param, mapObj);
  }
  return i18n.t(param);
};
