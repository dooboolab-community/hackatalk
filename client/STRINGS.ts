import * as Localization from 'expo-localization';

import en from './assets/langs/en.json';
import i18n from 'i18n-js';
import ko from './assets/langs/ko.json';

i18n.fallbacks = true;
i18n.translations = { en, ko };
i18n.locale = Localization.locale;

export const getString = (param: string, mapObj?: Record<string, unknown>): string => {
  if (mapObj) {
    return i18n.t(param, mapObj);
  }

  return i18n.t(param);
};
