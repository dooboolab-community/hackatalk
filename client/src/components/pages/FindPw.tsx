import {Button, EditText, useTheme} from 'dooboo-ui';
import React, {FC, useState} from 'react';
import type {
  UserFindPwMutation,
  UserFindPwMutationResponse,
} from '../../__generated__/UserFindPwMutation.graphql';
import {showAlertForError, validateEmail} from '../../utils/common';

import {Alert} from 'react-native';
import {AuthStackNavigationProps} from '../navigations/AuthStackNavigator';
import {PayloadError} from 'relay-runtime';
import {findPasswordMutation} from '../../relay/queries/User';
import {getString} from '../../../STRINGS';
import styled from '@emotion/native';
import {useMutation} from 'react-relay/hooks';
import {useNavigation} from '@react-navigation/core';

const Container = styled.View`
  flex: 1;
  padding: 0 32px;
  background-color: ${({theme}) => theme.background};
`;

const ButtonWrapper = styled.View`
  width: 100%;
  margin-top: 20px;
`;

const Page: FC = () => {
  const navigation = useNavigation<AuthStackNavigationProps<'FindPw'>>();
  const [email, setEmail] = useState<string>('');
  const [errorEmail, setErrorEmail] = useState<string>('');

  const [commitFindPassword, isInFlight] = useMutation<UserFindPwMutation>(
    findPasswordMutation,
  );

  const {theme} = useTheme();

  const onFindPw = async (): Promise<void> => {
    if (!validateEmail(email)) {
      setErrorEmail(getString('EMAIL_FORMAT_NOT_VALID'));

      return;
    }

    const mutationConfig = {
      variables: {email},
      onError: (error: Error) => {
        showAlertForError(error);
      },
      onCompleted: (
        response: UserFindPwMutationResponse,
        errors: PayloadError[] | null,
      ) => {
        const result = response.findPassword;

        if (errors && errors.length > 0) {
          Alert.alert(getString('FAILED'), getString('EMAIL_USER_NOT_EXISTS'));

          return;
        }

        if (result) {
          Alert.alert(
            getString('SUCCESS'),
            getString('PASSWORD_RESET_EMAIL_SENT'),
          );

          navigation.goBack();
        }
      },
    };

    commitFindPassword(mutationConfig);
  };

  return (
    <Container>
      <EditText
        testID="input-email"
        style={{marginVertical: 20}}
        styles={{
          container: {
            borderColor: theme.text,
          },
          input: {
            color: 'red',
          },
        }}
        focusColor={theme.focused}
        labelText={getString('EMAIL')}
        placeholderTextColor={theme.placeholder}
        placeholder="hello@example.com"
        value={email}
        onChangeText={(text: string): void => {
          setEmail(text);
          setErrorEmail('');
        }}
        errorText={errorEmail}
        onSubmitEditing={onFindPw}
      />
      <ButtonWrapper>
        <Button
          testID="btn-find-pw"
          loading={isInFlight}
          onPress={onFindPw}
          styles={{
            container: {
              height: 52,
              backgroundColor: theme.btnPrimary,
              borderWidth: 1,
            },
            text: {
              color: theme.btnPrimaryFont,
            },
            hovered: {
              borderColor: theme.text,
            },
          }}
          text={getString('FIND_PW')}
        />
      </ButtonWrapper>
    </Container>
  );
};

export default Page;
