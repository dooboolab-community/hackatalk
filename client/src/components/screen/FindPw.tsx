import { Button, EditText } from 'dooboo-ui';
import React, { ReactElement, useState } from 'react';
import { showAlertForError, validateEmail } from '../../utils/common';

import { Alert } from 'react-native';
import { AuthStackNavigationProps } from '../navigation/AuthStackNavigator';
import { getString } from '../../../STRINGS';
import styled from 'styled-components/native';
import { useThemeContext } from '@dooboo-ui/theme';

const Container = styled.View`
  flex: 1;
  background-color: ${({ theme }): string => theme.background};
`;

const Wrapper = styled.View`
  margin: 44px;
`;

const ButtonWrapper = styled.View`
  width: 100%;
  margin-top: 20px;
`;

interface Props {
  navigation: AuthStackNavigationProps<'FindPw'>;
}

function Page({ navigation }: Props): ReactElement {
  const [email, setEmail] = useState<string>('');
  const [errorEmail, setErrorEmail] = useState<string>('');
  const [findingPw, setFindingPw] = useState<boolean>(false);

  const { theme } = useThemeContext();
  // const [findPassword] = useMutation<{ findPassword: boolean }, MutationFindPasswordInput>(MUTATION_FIND_PASSWORD);

  const navigateToSignIn = (): void => navigation.navigate('SignIn');

  const onFindPw = async (): Promise<void> => {
    if (!validateEmail(email)) {
      setErrorEmail(getString('EMAIL_FORMAT_NOT_VALID'));
      return;
    }

    try {
      setFindingPw(true);
      // const result = await findPassword({ variables: { email } });
      // if (result.data?.findPassword) {
      //   Alert.alert('', getString('PASSWORD_RESET_EMAIL_SENT'), [
      //     {
      //       text: getString('OK'),
      //       onPress: navigateToSignIn,
      //     },
      //   ]);
      // }
    } catch ({ graphQLErrors }) {
      showAlertForError(graphQLErrors);
    } finally {
      setFindingPw(false);
    }
  };

  return (
    <Container>
      <Wrapper>
        <EditText
          testID="input-email"
          errorTestID="error-email"
          textStyle={{
            color: theme.fontColor,
          }}
          borderColor={theme.font}
          focusColor={theme.focused}
          placeholderTextColor={theme.placeholder}
          label={getString('EMAIL')}
          placeholder="hello@example.com"
          value={email}
          onChangeText={(text: string): void => {
            setEmail(text);
            setErrorEmail('');
          }}
          style={{ marginTop: 20 }}
          errorText={errorEmail}
          onSubmitEditing={onFindPw}
        />
        <ButtonWrapper>
          <Button
            testID="btn-find-pw"
            loading={findingPw}
            onPress={onFindPw}
            containerStyle={{ height: 52, backgroundColor: theme.btnPrimary }}
            textStyle={{ color: theme.btnPrimaryFont }}
            text={getString('FIND_PW')}
          />
        </ButtonWrapper>
      </Wrapper>
    </Container>
  );
}

export default Page;
