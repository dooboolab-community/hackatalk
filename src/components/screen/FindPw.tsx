import React, { ReactElement, useState } from 'react';

import { AuthStackNavigationProps } from '../navigation/AuthStackNavigator';
import Button from '../shared/Button';
import { EditText } from '@dooboo-ui/native';
import { getString } from '../../../STRINGS';
import styled from 'styled-components/native';
import { useThemeContext } from '@dooboo-ui/native-theme';
import { validateEmail } from '../../utils/common';

const Container = styled.View`
  flex: 1;
  background-color: ${({ theme }): string => theme.background};
`;

const Wrapper = styled.View`
  margin: 44px;
`;

const ButtonWrapper = styled.View`
  width: 100%;
  margin-top: 20px;
`;

interface Props {
  navigation: AuthStackNavigationProps<'FindPw'>;
}

function Page(props: Props): ReactElement {
  const [email, setEmail] = useState<string>('');
  const [errorEmail, setErrorEmail] = useState<string>('');
  const [findingPw, setFindingPw] = useState<boolean>(false);
  let timer: number;

  const { theme } = useThemeContext();

  const onFindPw = (): void => {
    if (!validateEmail(email)) {
      setErrorEmail(getString('EMAIL_FORMAT_NOT_VALID'));
      return;
    }

    setFindingPw(true);
    timer = setTimeout(() => {
      setFindingPw(false);
      clearTimeout(timer);
      if (props.navigation) {
        props.navigation.navigate('SignIn');
      }
    }, 1000);
  };

  return (
    <Container>
      <Wrapper>
        <EditText
          testID="input-email"
          errorTestID="error-email"
          textStyle={{
            color: theme.fontColor,
          }}
          label={getString('EMAIL')}
          placeholder="hello@example.com"
          placeholderTextColor={theme.placeholder}
          value={email}
          onChangeText={(text: string): void => {
            setEmail(text);
            setErrorEmail('');
          }}
          style={{ marginTop: 20 }}
          errorText={errorEmail}
          onSubmitEditing={onFindPw}
        />
        <ButtonWrapper>
          <Button
            testID="btn-find-pw"
            isLoading={findingPw}
            onPress={onFindPw}
            containerStyle={{ padding: 5, height: 52 }}
          >
            {getString('FIND_PW')}
          </Button>
        </ButtonWrapper>
      </Wrapper>
    </Container>
  );
}

export default Page;
