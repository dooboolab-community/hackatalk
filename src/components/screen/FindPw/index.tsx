import * as yup from 'yup';

import { Container, ErrorText, InnerContainer, StyledButtonWrapper, StyledScrollView } from './styles';
import { NavigationParams, NavigationScreenProp, NavigationState } from 'react-navigation';
import React, { ReactElement, memo, useCallback } from 'react';

import { Alert } from 'react-native';
import Button from '../../shared/Button';
import TextInput from '../../shared/TextInput';
import { getString } from '../../../../STRINGS';
import useForm from 'react-hook-form';

interface Props {
  navigation?: NavigationScreenProp<NavigationState, NavigationParams>;
}

const schema = yup.object().shape({
  email: yup
    .string()
    .email(getString('EMAIL_FORMAT_NOT_VALID'))
    .required(getString('EMAIL_REQUIRED')),
});

function FindPw(): ReactElement {
  const {
    register,
    errors,
    setValue,
    handleSubmit,
    formState: { touched },
    watch,
    reset,
  } = useForm({
    validationSchema: schema,
  });
  const onSubmit = useCallback(({ email }) => {
    Alert.alert(
      'Find Password',
      `password reset link has been sent to your email: ${email}.`,
    );
    reset({ email: '' });
  }, []);
  const onTextChanged = useCallback((text: string): void | Promise<boolean> => {
    setValue('email', text, true);
  }, []);
  const errorMessage = errors.email && errors.email.message;

  return (
    <Container>
      <StyledScrollView>
        <InnerContainer>
          <TextInput
            testID="findPw_email_input"
            style={{ marginTop: 8 }}
            txtLabel={getString('EMAIL')}
            txtHint={getString('EMAIL')}
            txt={watch('email', '')}
            ref={register({ name: 'email' })}
            onTextChanged={onTextChanged}
            error={errorMessage}
          />
          {!!errors.email && (
            <ErrorText isError={!!errors.email}>
              {errors.email.message}
            </ErrorText>
          )}
          <StyledButtonWrapper>
            <Button
              testID="btnFindPwConfirm"
              isDisabled={!(touched && touched.length) || !!errorMessage}
              onPress={handleSubmit(onSubmit)}
              width="100%"
              containerStyle={{ flexDirection: 'row' }}
            >
              {getString('PASSWORD_RESET')}
            </Button>
          </StyledButtonWrapper>
        </InnerContainer>
      </StyledScrollView>
    </Container>
  );
}

export default memo(FindPw);
