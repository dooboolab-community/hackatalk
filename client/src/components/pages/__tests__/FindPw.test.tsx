import {MockPayloadGenerator, createMockEnvironment} from 'relay-test-utils';
import {
  createMockNavigation,
  createTestElement,
} from '../../../../test/testUtils';
import {fireEvent, render, waitFor} from '@testing-library/react-native';

import {Alert} from 'react-native';
import FindPw from '../FindPw';
import React from 'react';
import ReactNavigation from '@react-navigation/core';
import {getString} from '../../../../STRINGS';

const mockNavigation = createMockNavigation();

jest.mock('@react-navigation/core', () => ({
  ...jest.requireActual<typeof ReactNavigation>('@react-navigation/core'),
  useNavigation: () => mockNavigation,
}));

describe('[FindPw] rendering test', () => {
  it('should renders as expected', () => {
    const component = createTestElement(<FindPw />);
    const screen = render(component);
    const json = screen.toJSON();

    expect(json).toBeTruthy();
    expect(json).toMatchSnapshot();
  });
});

describe('[FindPw] interaction', () => {
  it('should invoke changeText event handler when email changed', () => {
    const component = createTestElement(<FindPw />);
    const screen = render(component);
    const textInput = screen.getByTestId('input-email');

    fireEvent.changeText(textInput, 'email@email.com');

    expect(textInput.props.value).toEqual('email@email.com');
  });

  describe('onFindPw', () => {
    it('should show error text when the email is not validated', () => {
      const component = createTestElement(<FindPw />);
      const screen = render(component);
      const textInput = screen.getByTestId('input-email');

      fireEvent.changeText(textInput, 'example');

      const btnFindPw = screen.getByTestId('btn-find-pw');

      fireEvent.press(btnFindPw);

      const errorText = screen.getByText(getString('EMAIL_FORMAT_NOT_VALID'));

      expect(errorText).toBeTruthy();
    });

    it('should call FindPw when button has clicked and navigate to SignIn', async () => {
      const mockEnvironment = createMockEnvironment();

      mockEnvironment.mock.queueOperationResolver((operation) =>
        MockPayloadGenerator.generate(operation),
      );

      const mockAlert = jest.spyOn(Alert, 'alert');

      const component = createTestElement(<FindPw />, {
        environment: mockEnvironment,
      });

      const screen = render(component);
      const textInput = screen.getByTestId('input-email');

      fireEvent.changeText(textInput, 'email@email.com');

      const btnFindPw = screen.getByTestId('btn-find-pw');

      fireEvent.press(btnFindPw);

      await waitFor(() => expect(mockAlert).toHaveBeenCalled());

      expect(mockAlert).toHaveBeenCalledWith(
        expect.any(String),
        getString('PASSWORD_RESET_EMAIL_SENT'),
      );
    });
  });
});
