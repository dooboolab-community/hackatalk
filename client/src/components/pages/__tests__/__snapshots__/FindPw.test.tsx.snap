// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[FindPw] rendering test should renders as expected 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#000",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "paddingBottom": 0,
          "paddingLeft": 32,
          "paddingRight": 32,
          "paddingTop": 0,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignSelf": "stretch",
            "flexDirection": "column",
          },
          Object {
            "marginBottom": 20,
            "marginTop": 32,
          },
        ]
      }
      testID="container-id"
    >
      <View
        style={
          Array [
            Array [
              Object {
                "alignSelf": "stretch",
                "borderBottomColor": "white",
                "borderBottomWidth": 1,
                "flexDirection": "column",
                "justifyContent": "space-between",
              },
              Object {
                "borderColor": "white",
              },
            ],
            false,
            Object {
              "borderColor": "white",
            },
            Object {
              "borderBottomColor": "white",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Array [
                Object {
                  "color": "#6D6D6D",
                  "fontSize": 14,
                  "paddingHorizontal": 4,
                },
                undefined,
              ],
              false,
              Object {},
            ]
          }
        >
          Email
        </Text>
        <TextInput
          allowFontScaling={true}
          autoCapitalize="none"
          editable={true}
          multiline={false}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          onSubmitEditing={[Function]}
          placeholder="hello@example.com"
          placeholderTextColor="#6D6D6D"
          rejectResponderTermination={true}
          secureTextEntry={false}
          style={
            Array [
              Array [
                Object {
                  "color": "white",
                  "fontSize": 14,
                  "paddingHorizontal": 10,
                  "paddingVertical": 10,
                },
                Object {
                  "color": "white",
                },
              ],
              false,
            ]
          }
          testID="input-email"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          style={
            Array [
              Array [
                Object {
                  "color": "#FF002E",
                  "fontSize": 12,
                  "marginTop": 8,
                  "paddingHorizontal": 10,
                },
                undefined,
              ],
              Object {
                "color": "#FF002E",
                "flex": 1,
              },
            ]
          }
        >
          
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "width": "100%",
          },
          undefined,
        ]
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="btn-find-pw"
      />
    </View>
  </View>
</RNCSafeAreaProvider>
`;
